proxy/http2/Http2Stream.cc

```diff
@@ -124,6 +124,9 @@ Http2Stream::main_event_handler(int event, void *edata)
   reentrancy_count--;
   // Clean stream up if the terminate flag is set and we are at the bottom of the handler stack
   if (terminate_stream && reentrancy_count == 0) {
+    Http2ClientSession *h2_parent = static_cast<Http2ClientSession *>(parent);
+    SCOPED_MUTEX_LOCK(lock, h2_parent->connection_state.mutex, this_ethread());
+    h2_parent->connection_state.delete_stream(this);
```

Comment:
These block looks similar to L667-L681 ( `Http2Stream::destroy()` ).
https://github.com/apache/trafficserver/blob/6dcc822baf23b1ab69c9a2609fc4f65fd9f6c10b/proxy/http2/Http2Stream.cc#L667-L681
How about calling `delete_stream()` before `release_stream()` in `Http2Stream::destroy()` ?

Reply:
I think the block in `destroy()` is a part of safety net. That may strengthen the safety net, however, I think `destroy` should do only destroying, and we shouldn't depend on the safety net.

Reply:
Yes, I agree that the delete_stream/release_stream/destroy relationship isn't very satisfying.  I'll look a bit, and see if we can unify things a bit.

Reply:
I have a different fix that I'm running on a prod box for 30 minutes (a new record!).  I made a change in release_stream.  It was removing the stream from the connection_state.  This means that the actual delete_stream would exit immediately without cleaning up the priority tree.  

I must run out right now, but I'll put up another PR later this evening. Assuming the prod box doesn't go crazy before.

Reply:
My fix is in PR #2781